File: app/src/main/java/com/elementary/tasks/core/app_widgets/calendar/CalendarThemeFragment.java
package com.elementary.tasks.core.app_widgets.calendar;

import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.elementary.tasks.R;
import com.elementary.tasks.core.calendar.FlextHelper;
import com.elementary.tasks.core.utils.Prefs;
import com.elementary.tasks.core.utils.ThemeUtil;
import com.elementary.tasks.databinding.FragmentCalendarWidgetPreviewBinding;
import com.elementary.tasks.navigation.fragments.BaseNavigationFragment;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.Formatter;
import java.util.GregorianCalendar;
import java.util.List;
import java.util.Locale;

import hirondelle.date4j.DateTime;

/**
 * Copyright 2015 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class CalendarThemeFragment extends BaseNavigationFragment {

    static final String ARGUMENT_PAGE_NUMBER = "arg_page_number";
    static final String ARGUMENT_DATA = "arg_data";
    private int pageNumber;
    private List<CalendarTheme> list;

    private ThemeUtil themeUtil;

    public static CalendarThemeFragment newInstance(int page, List<CalendarTheme> list) {
        CalendarThemeFragment pageFragment = new CalendarThemeFragment();
        Bundle arguments = new Bundle();
        arguments.putInt(ARGUMENT_PAGE_NUMBER, page);
        arguments.putParcelableArrayList(ARGUMENT_DATA, new ArrayList<>(list));
        pageFragment.setArguments(arguments);
        return pageFragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle intent = getArguments();
        pageNumber = intent.getInt(ARGUMENT_PAGE_NUMBER);
        list = intent.getParcelableArrayList(ARGUMENT_DATA);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        FragmentCalendarWidgetPreviewBinding binding = FragmentCalendarWidgetPreviewBinding.inflate(inflater, container, false);
        themeUtil = ThemeUtil.getInstance(getContext());
        CalendarTheme calendarTheme = list.get(pageNumber);
        int windowColor = calendarTheme.getWindowColor();
        binding.previewView.widgetBg.setBackgroundResource(windowColor);
        int windowTextColor = calendarTheme.getWindowTextColor();
        binding.themeTitle.setTextColor(windowTextColor);
        binding.note.setTextColor(windowTextColor);

        int itemTextColor = calendarTheme.getItemTextColor();
        int widgetBgColor = calendarTheme.getWidgetBgColor();
        int headerColor = calendarTheme.getHeaderColor();
        int borderColor = calendarTheme.getBorderColor();
        int titleColor = calendarTheme.getTitleColor();
        int rowColor = calendarTheme.getRowColor();

        int leftArrow = calendarTheme.getLeftArrow();
        int rightArrow = calendarTheme.getRightArrow();
        int iconPlus = calendarTheme.getIconPlus();
        int iconVoice = calendarTheme.getIconVoice();
        int iconSettings = calendarTheme.getIconSettings();

        int currentMark = calendarTheme.getCurrentMark();
        int birthdayMark = calendarTheme.getBirthdayMark();
        int reminderMark = calendarTheme.getReminderMark();

        binding.previewView.weekdayGrid.setBackgroundResource(widgetBgColor);
        binding.previewView.header.setBackgroundResource(headerColor);
        binding.previewView.currentDate.setTextColor(titleColor);
        binding.previewView.monthGrid.setBackgroundResource(borderColor);

        binding.previewView.plusButton.setImageResource(iconPlus);
        binding.previewView.nextMonth.setImageResource(rightArrow);
        binding.previewView.prevMonth.setImageResource(leftArrow);
        binding.previewView.voiceButton.setImageResource(iconVoice);
        binding.previewView.settingsButton.setImageResource(iconSettings);

        StringBuilder monthYearStringBuilder = new StringBuilder(50);
        Formatter monthYearFormatter = new Formatter(monthYearStringBuilder, Locale.getDefault());
        int monthYearFlag = DateUtils.FORMAT_SHOW_DATE | DateUtils.FORMAT_NO_MONTH_DAY | DateUtils.FORMAT_SHOW_YEAR;
        Calendar cal = new GregorianCalendar();
        String monthTitle = DateUtils.formatDateRange(getActivity(),
                monthYearFormatter, cal.getTimeInMillis(), cal.getTimeInMillis(), monthYearFlag).toString();
        binding.previewView.currentDate.setText(monthTitle.toUpperCase());

        binding.themeTitle.setText(calendarTheme.getTitle());

        binding.previewView.weekdayGrid.setAdapter(new WeekdayAdapter(getActivity(), itemTextColor));
        binding.previewView.monthGrid.setAdapter(new MonthGridAdapter(getActivity(), new int[]{itemTextColor, rowColor,
                currentMark, birthdayMark, reminderMark}));
        return binding.getRoot();
    }

    private class WeekdayAdapter extends BaseAdapter{

        private List<String> weekdays;
        private int SUNDAY = 1;
        private int startDayOfWeek = SUNDAY;
        private Context context;
        private LayoutInflater inflater;
        private int textColor;

        WeekdayAdapter(Context context, int textColor){
            this.context = context;
            this.textColor = textColor;
            inflater = LayoutInflater.from(context);
            weekdays = new ArrayList<>();
            weekdays.clear();
            SimpleDateFormat fmt = new SimpleDateFormat("EEE", Locale.getDefault());

            // 17 Feb 2013 is Sunday
            DateTime sunday = new DateTime(2013, 2, 17, 0, 0, 0, 0);
            DateTime nextDay = sunday.plusDays(startDayOfWeek - SUNDAY);
            if (getPrefs().getStartDay() == 1){
                nextDay = nextDay.plusDays(1);
            }
            for (int i = 0; i < 7; i++) {
                Date date = FlextHelper.convertDateTimeToDate(nextDay);
                weekdays.add(fmt.format(date).toUpperCase());
                nextDay = nextDay.plusDays(1);
            }
        }

        @Override
        public int getCount() {
            return weekdays.size();
        }

        @Override
        public Object getItem(int position) {
            return weekdays.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            if (convertView == null) {
                inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
                convertView = inflater.inflate(R.layout.weekday_grid, null);
            }
            TextView textView = (TextView) convertView.findViewById(R.id.textView1);
            textView.setText(weekdays.get(position));
            textView.setTextColor(textColor);
            return convertView;
        }

        @Override
        public int getViewTypeCount() {
            return 1;
        }

        @Override
        public boolean hasStableIds() {
            return true;
        }
    }

    private class MonthGridAdapter extends BaseAdapter{

        List<DateTime> datetimeList;
        int SUNDAY = 1;
        int startDayOfWeek = SUNDAY;
        int prefsMonth;
        Context context;
        LayoutInflater inflater;
        int textColor;
        int widgetBgColor;
        int cMark;
        int bMark;
        int rMark;

        MonthGridAdapter(Context context, int[] resources){
            this.context = context;
            this.textColor = resources[0];
            this.widgetBgColor = resources[1];
            this.cMark = resources[2];
            this.bMark = resources[3];
            this.rMark = resources[4];
            inflater = LayoutInflater.from(context);
            datetimeList = new ArrayList<>();
            datetimeList.clear();

            Calendar calendar = Calendar.getInstance();
            calendar.setTimeInMillis(System.currentTimeMillis());
            int year = calendar.get(Calendar.YEAR);

            DateTime firstDateOfMonth = new DateTime(year, prefsMonth + 1, 1, 0, 0, 0, 0);
            DateTime lastDateOfMonth = firstDateOfMonth.plusDays(firstDateOfMonth.getNumDaysInMonth() - 1);

            int weekdayOfFirstDate = firstDateOfMonth.getWeekDay();
            if (weekdayOfFirstDate < startDayOfWeek) {
                weekdayOfFirstDate += 7;
            }
            while (weekdayOfFirstDate > 0) {
                int temp = startDayOfWeek;
                if (Prefs.getInstance(context).getStartDay() == 1){
                    temp = startDayOfWeek + 1;
                }
                DateTime dateTime = firstDateOfMonth.minusDays(weekdayOfFirstDate - temp);
                if (!dateTime.lt(firstDateOfMonth)) {
                    break;
                }
                datetimeList.add(dateTime);
                weekdayOfFirstDate--;
            }
            for (int i = 0; i < lastDateOfMonth.getDay(); i++) {
                datetimeList.add(firstDateOfMonth.plusDays(i));
            }
            int endDayOfWeek = startDayOfWeek - 1;
            if (endDayOfWeek == 0) {
                endDayOfWeek = 7;
            }
            if (lastDateOfMonth.getWeekDay() != endDayOfWeek) {
                int i = 1;
                while (true) {
                    DateTime nextDay = lastDateOfMonth.plusDays(i);
                    datetimeList.add(nextDay);
                    i++;
                    if (nextDay.getWeekDay() == endDayOfWeek) {
                        break;
                    }
                }
            }
            int size = datetimeList.size();
            int numOfDays = 42 - size;
            DateTime lastDateTime = datetimeList.get(size - 1);
            for (int i = 1; i <= numOfDays; i++) {
                DateTime nextDateTime = lastDateTime.plusDays(i);
                datetimeList.add(nextDateTime);
            }
        }

        @Override
        public int getCount() {
            return datetimeList.size();
        }

        @Override
        public Object getItem(int position) {
            return datetimeList.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            if (convertView == null) {
                inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
                convertView = inflater.inflate(R.layout.month_view_grid, null);
            }
            int selDay = datetimeList.get(position).getDay();
            FrameLayout background = (FrameLayout) convertView.findViewById(R.id.background);
            TextView textView = (TextView) convertView.findViewById(R.id.textView);
            TextView currentMark = (TextView) convertView.findViewById(R.id.currentMark);
            TextView reminderMark = (TextView) convertView.findViewById(R.id.reminderMark);
            TextView birthdayMark = (TextView) convertView.findViewById(R.id.birthdayMark);

            textView.setText(String.valueOf(selDay));
            textView.setTextColor(textColor);
            background.setBackgroundResource(widgetBgColor);

            currentMark.setBackgroundColor(Color.TRANSPARENT);
            reminderMark.setBackgroundColor(Color.TRANSPARENT);
            birthdayMark.setBackgroundColor(Color.TRANSPARENT);
            if (selDay == 15){
                if (rMark != 0){
                    reminderMark.setBackgroundResource(rMark);
                } else {
                    reminderMark.setBackgroundColor(context.getResources()
                            .getColor(themeUtil.colorReminderCalendar()));
                }
            }
            if (selDay == 11){
                if (bMark != 0){
                    birthdayMark.setBackgroundResource(bMark);
                } else {
                    birthdayMark.setBackgroundColor(context.getResources()
                            .getColor(themeUtil.colorBirthdayCalendar()));
                }
            }
            if (11 == selDay){
                if (cMark != 0){
                    currentMark.setBackgroundResource(cMark);
                } else {
                    currentMark.setBackgroundColor(context.getResources()
                            .getColor(themeUtil.colorCurrentCalendar()));
                }
            }
            return convertView;
        }

        @Override
        public int getViewTypeCount() {
            return 1;
        }

        @Override
        public boolean hasStableIds() {
            return true;
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/appWidgets/calendar/CalendarThemeFragment.kt
package com.elementary.tasks.core.appWidgets.calendar

import android.content.Context
import android.graphics.Color
import android.os.BadParcelableException
import android.os.Bundle
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.FrameLayout
import android.widget.TextView
import com.elementary.tasks.R
import com.elementary.tasks.core.calendar.FlextHelper
import com.elementary.tasks.navigation.fragments.BaseNavigationFragment
import hirondelle.date4j.DateTime
import kotlinx.android.synthetic.main.fragment_calendar_widget_preview.*
import kotlinx.android.synthetic.main.widget_calendar.view.*
import java.text.SimpleDateFormat
import java.util.*

/**
 * Copyright 2015 Nazar Suhovich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class CalendarThemeFragment : BaseNavigationFragment() {
    private var pageNumber: Int = 0
    private var list: List<CalendarTheme> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val intent = arguments
        try {
            pageNumber = intent?.getInt(ARGUMENT_PAGE_NUMBER) ?: 0
            val list = intent?.getParcelableArrayList<CalendarTheme>(ARGUMENT_DATA)
            if (list != null) this.list = list
        } catch (e: BadParcelableException) {
        }
    }

    override fun layoutRes(): Int = R.layout.fragment_calendar_widget_preview

    override fun getTitle(): String = ""

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val calendarTheme = list[pageNumber]
        val windowColor = calendarTheme.windowColor
        previewView.widgetBg.setBackgroundResource(windowColor)
        val windowTextColor = calendarTheme.windowTextColor
        themeTitle.setTextColor(windowTextColor)
        note.setTextColor(windowTextColor)

        val itemTextColor = calendarTheme.itemTextColor
        val widgetBgColor = calendarTheme.widgetBgColor
        val headerColor = calendarTheme.headerColor
        val borderColor = calendarTheme.borderColor
        val titleColor = calendarTheme.titleColor
        val rowColor = calendarTheme.rowColor

        val leftArrow = calendarTheme.leftArrow
        val rightArrow = calendarTheme.rightArrow
        val iconPlus = calendarTheme.iconPlus
        val iconVoice = calendarTheme.iconVoice
        val iconSettings = calendarTheme.iconSettings

        val currentMark = calendarTheme.currentMark
        val birthdayMark = calendarTheme.birthdayMark
        val reminderMark = calendarTheme.reminderMark

        previewView.weekdayGrid.setBackgroundResource(widgetBgColor)
        previewView.header.setBackgroundResource(headerColor)
        previewView.currentDate.setTextColor(titleColor)
        previewView.monthGrid.setBackgroundResource(borderColor)

        previewView.plusButton.setImageResource(iconPlus)
        previewView.nextMonth.setImageResource(rightArrow)
        previewView.prevMonth.setImageResource(leftArrow)
        previewView.voiceButton.setImageResource(iconVoice)
        previewView.settingsButton.setImageResource(iconSettings)

        val monthYearStringBuilder = StringBuilder(50)
        val monthYearFormatter = Formatter(monthYearStringBuilder, Locale.getDefault())
        val monthYearFlag = DateUtils.FORMAT_SHOW_DATE or DateUtils.FORMAT_NO_MONTH_DAY or DateUtils.FORMAT_SHOW_YEAR
        val cal = GregorianCalendar()
        val monthTitle = DateUtils.formatDateRange(activity,
                monthYearFormatter, cal.timeInMillis, cal.timeInMillis, monthYearFlag).toString()
        previewView.currentDate.text = monthTitle.toUpperCase()

        themeTitle.text = calendarTheme.title

        previewView.weekdayGrid.adapter = WeekdayAdapter(activity!!, itemTextColor)
        previewView.monthGrid.adapter = MonthGridAdapter(activity!!, intArrayOf(itemTextColor, rowColor, currentMark, birthdayMark, reminderMark))
    }

    private inner class WeekdayAdapter internal constructor(private val context: Context, private val textColor: Int) : BaseAdapter() {

        private val weekdays: MutableList<String>
        private val sunday = 1
        private val startDayOfWeek = sunday
        private var inflater: LayoutInflater? = null

        init {
            inflater = LayoutInflater.from(context)
            weekdays = ArrayList()
            weekdays.clear()
            val fmt = SimpleDateFormat("EEE", Locale.getDefault())

            // 17 Feb 2013 is Sunday
            val sunday = DateTime(2013, 2, 17, 0, 0, 0, 0)
            var nextDay = sunday.plusDays(startDayOfWeek - this.sunday)
            if (prefs!!.startDay == 1) {
                nextDay = nextDay.plusDays(1)
            }
            for (i in 0..6) {
                val date = FlextHelper.convertDateTimeToDate(nextDay)
                weekdays.add(fmt.format(date).toUpperCase())
                nextDay = nextDay.plusDays(1)
            }
        }

        override fun getCount(): Int {
            return weekdays.size
        }

        override fun getItem(position: Int): Any {
            return weekdays[position]
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            var cView = convertView
            if (cView == null) {
                inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
                cView = inflater!!.inflate(R.layout.list_item_weekday_grid, null)
            }
            val textView = cView!!.findViewById<TextView>(R.id.textView1)
            textView.text = weekdays[position]
            textView.setTextColor(textColor)
            return cView
        }

        override fun getViewTypeCount(): Int {
            return 1
        }

        override fun hasStableIds(): Boolean {
            return true
        }
    }

    private inner class MonthGridAdapter internal constructor(internal var context: Context, resources: IntArray) : BaseAdapter() {

        internal var datetimeList: MutableList<DateTime>
        internal var startDayOfWeek = 1
        internal var prefsMonth: Int = 0
        internal var inflater: LayoutInflater
        internal var textColor: Int = 0
        internal var widgetBgColor: Int = 0
        internal var cMark: Int = 0
        internal var bMark: Int = 0
        internal var rMark: Int = 0

        init {
            this.textColor = resources[0]
            this.widgetBgColor = resources[1]
            this.cMark = resources[2]
            this.bMark = resources[3]
            this.rMark = resources[4]
            inflater = LayoutInflater.from(context)
            datetimeList = ArrayList()
            datetimeList.clear()

            val calendar = Calendar.getInstance()
            calendar.timeInMillis = System.currentTimeMillis()
            val year = calendar.get(Calendar.YEAR)

            val firstDateOfMonth = DateTime(year, prefsMonth + 1, 1, 0, 0, 0, 0)
            val lastDateOfMonth = firstDateOfMonth.plusDays(firstDateOfMonth.numDaysInMonth - 1)

            var weekdayOfFirstDate = firstDateOfMonth.weekDay!!
            if (weekdayOfFirstDate < startDayOfWeek) {
                weekdayOfFirstDate += 7
            }
            while (weekdayOfFirstDate > 0) {
                var temp = startDayOfWeek
                if (prefs.startDay == 1) {
                    temp = startDayOfWeek + 1
                }
                val dateTime = firstDateOfMonth.minusDays(weekdayOfFirstDate - temp)
                if (!dateTime.lt(firstDateOfMonth)) {
                    break
                }
                datetimeList.add(dateTime)
                weekdayOfFirstDate--
            }
            for (i in 0 until lastDateOfMonth.day) {
                datetimeList.add(firstDateOfMonth.plusDays(i))
            }
            var endDayOfWeek = startDayOfWeek - 1
            if (endDayOfWeek == 0) {
                endDayOfWeek = 7
            }
            if (lastDateOfMonth.weekDay != endDayOfWeek) {
                var i = 1
                while (true) {
                    val nextDay = lastDateOfMonth.plusDays(i)
                    datetimeList.add(nextDay)
                    i++
                    if (nextDay.weekDay == endDayOfWeek) {
                        break
                    }
                }
            }
            val size = datetimeList.size
            val numOfDays = 42 - size
            val lastDateTime = datetimeList[size - 1]
            for (i in 1..numOfDays) {
                val nextDateTime = lastDateTime.plusDays(i)
                datetimeList.add(nextDateTime)
            }
        }

        override fun getCount(): Int {
            return datetimeList.size
        }

        override fun getItem(position: Int): Any {
            return datetimeList[position]
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            var cView = convertView
            if (cView == null) {
                inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
                cView = inflater.inflate(R.layout.view_month_grid, null)
            }
            val selDay = datetimeList[position].day!!
            val background = cView!!.findViewById<FrameLayout>(R.id.background)
            val textView = cView.findViewById<TextView>(R.id.textView)
            val currentMark = cView.findViewById<TextView>(R.id.currentMark)
            val reminderMark = cView.findViewById<TextView>(R.id.reminderMark)
            val birthdayMark = cView.findViewById<TextView>(R.id.birthdayMark)

            textView.text = selDay.toString()
            textView.setTextColor(textColor)
            background.setBackgroundResource(widgetBgColor)

            currentMark.setBackgroundColor(Color.TRANSPARENT)
            reminderMark.setBackgroundColor(Color.TRANSPARENT)
            birthdayMark.setBackgroundColor(Color.TRANSPARENT)
            if (selDay == 15) {
                if (rMark != 0) {
                    reminderMark.setBackgroundResource(rMark)
                } else {
                    reminderMark.setBackgroundColor(context.resources
                            .getColor(themeUtil!!.colorReminderCalendar()))
                }
            }
            if (selDay == 11) {
                if (bMark != 0) {
                    birthdayMark.setBackgroundResource(bMark)
                } else {
                    birthdayMark.setBackgroundColor(context.resources
                            .getColor(themeUtil!!.colorBirthdayCalendar()))
                }
            }
            if (11 == selDay) {
                if (cMark != 0) {
                    currentMark.setBackgroundResource(cMark)
                } else {
                    currentMark.setBackgroundColor(context.resources
                            .getColor(themeUtil!!.colorCurrentCalendar()))
                }
            }
            return cView
        }

        override fun getViewTypeCount(): Int {
            return 1
        }

        override fun hasStableIds(): Boolean {
            return true
        }
    }

    companion object {

        internal const val ARGUMENT_PAGE_NUMBER = "arg_page_number"
        internal const val ARGUMENT_DATA = "arg_data"

        fun newInstance(page: Int, list: List<CalendarTheme>): CalendarThemeFragment {
            val pageFragment = CalendarThemeFragment()
            val arguments = Bundle()
            arguments.putInt(ARGUMENT_PAGE_NUMBER, page)
            arguments.putParcelableArrayList(ARGUMENT_DATA, ArrayList(list))
            pageFragment.arguments = arguments
            return pageFragment
        }
    }
}
